maro/simulator/scenarios/data_center/business_engine.py

```diff
@@ -0,0 +1,253 @@
+                vm = self._vm[vm_id]
+                cur_util_cores = vm.util_cpu * vm.req_cpu / 100
+                total_cpu += cur_util_cores
+            pm.util_cpu = total_cpu / pm.cap_cpu * 100
+            pm.update_util_series(self._tick)
+
+        # TODO: Energy comsumption update.
+
+    def _on_vm_required(self, evt: CascadeEvent):
+        """Callback when there is a VM requirement generated."""
+        # Get VM data from payload.
+        payload: VmRequirementPayload = evt.payload
+        vm_req: VirtualMachine = payload.vm_req
+        buffer_time: int = payload.buffer_time
+
+        # Check all valid PMs.
+        # NOTE: Should we implement this logic inside the action scope?
+        valid_pm_list = [
+            pm.id
+            for pm in self._machines
+            if (pm.cap_cpu - (pm.cap_cpu * pm.util_cpu / 100)) >= vm_req.req_cpu
+        ]
+
+        if len(valid_pm_list) > 0:
+            # Generate pending decision.
+            decision_payload = DecisionEvent(
+                valid_pm=valid_pm_list,
+                vm_info=vm_req,
+                buffer_time=buffer_time
+            )
+            pending_decision_evt = self._event_buffer.gen_decision_event(evt.tick, payload=decision_payload)
+            evt.add_immediate_event(pending_decision_evt)
+            self._pending_action_vm_id = vm_req.vm_id
+        else:
+            # Postpone the buffer duration ticks by config setting.
+            if buffer_time > 0:
+                postpone_payload = payload
+                postpone_payload.buffer_time -= self._delay_duration
+                self._total_latency.latency_due_to_resource += self._delay_duration
+                postpone_evt = self._event_buffer.gen_cascade_event(
+                    evt.tick + self._delay_duration, Events.REQUIREMENTS, payload=postpone_payload)
+                self._event_buffer.insert_event(postpone_evt)
+            else:
+                # Fail
+                # TODO Implement failure logic.
+                self._failed_requirements += 1
+
+    def _on_vm_finished(self, evt: AtomEvent):
+        """Callback when there is a VM in the end cycle."""
+        # Get the end-cycle VM info.
+        payload: VmFinishedPayload = evt.payload
+        vm_id = payload.vm_id
+        virtual_machine: VirtualMachine = self._vm[vm_id]
+        util_vm_cores = virtual_machine.util_cpu * virtual_machine.req_cpu
+
+        # Release PM resources.
+        physical_machine: PhysicalMachine = self._machines[virtual_machine.pm_id]
+        physical_machine.req_cpu -= virtual_machine.req_cpu
+        physical_machine.req_mem -= virtual_machine.req_mem
+        physical_machine.util_cpu = (
+            (physical_machine.cap_cpu * physical_machine.util_cpu / 100 - util_vm_cores)
+            * 100 / physical_machine.cap_cpu
+        )
+        physical_machine.remove_vm(vm_id)
+
+        # Remove dead VM.
+        self._vm.pop(vm_id)
+
+        # VM allocation succeed.
```

Comment:
We should unify the wording in code, docstrings, comments, and also review-comments.

e.g.
subscription v.s. allocation v.s. assign ...